src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -1003,7 +1021,6 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
                     // Be conservative about the returned value of a const fn.
                     let tcx = self.tcx;
                     let ty = dest.ty(self.mir, tcx).to_ty(tcx);
-                    self.qualif = Qualif::empty();
```

Comment:
Why did you remove this line?

Reply:
Otherwise we'd be erasing the `NOT_PROMOTABLE` qualification we added above.

Reply:
Yep, but I think you want to reset all fields apart from that one, no? Alternatively, just make sure that's set after `self.qualif` is reset.